File: Javascript/Linked List/面试题 02.01. 移除重复节点.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
/*
 * 用q表示p的父节点
 * 如果该节点已经存在则将p = p.next q.next = p
*/ 
var removeDuplicateNodes = function(head) {
    let q;
    let p = head;
    let obj = {}
    while(p){
        if(!obj[p.val]) {
            obj[p.val] = true
            p=p.next
            if(!q) {
                q= head
            }else{
                q = q.next
            }
        }else{
            p=p.next
            q.next = p
        } 
    }
    return head
};

File: Go/LinkedList/面试题 02.01. 移除重复节点.go
package LinkedList

// 用map记录节点数值
func removeDuplicateNodes(head *ListNode) *ListNode {
	repeat := make(map[int]bool)
	res := head
	// 头部添加一个额外节点
	head = &ListNode{
		Next: head,
	}
	for head != nil && head.Next != nil {
		if _, ok := repeat[head.Next.Val]; !ok {
			repeat[head.Next.Val] = true
			head = head.Next
		} else {
			head.Next = head.Next.Next
		}
	}
	printLL(res)
	return res
}